src/serac/numerics/tests/equationsolver_petsc.cpp

```diff
@@ -0,0 +1,171 @@
+  auto pmesh = mfem::ParMesh(MPI_COMM_WORLD, mesh);
+
+  pmesh.EnsureNodes();
+  pmesh.ExchangeFaceNbrData();
+
+  constexpr int p   = 1;
+  constexpr int dim = 2;
+
+  // Create standard MFEM bilinear and linear forms on H1
+  auto                        fec = mfem::H1_FECollection(p, dim);
+  mfem::ParFiniteElementSpace fes(&pmesh, &fec);
+
+  mfem::HypreParVector x_exact(&fes);
+  mfem::HypreParVector x_computed(&fes);
+
+  std::unique_ptr<mfem::HypreParMatrix> J;
+
+  // Define the types for the test and trial spaces using the function arguments
+  using test_space  = H1<p>;
+  using trial_space = H1<p>;
+
+  // Construct the new functional object using the known test and trial spaces
+  Functional<test_space(trial_space)> residual(&fes, {&fes});
+
+  x_exact.Randomize(0);
+
+  residual.AddDomainIntegral(
+      Dimension<dim>{}, DependsOn<0>{},
+      [&](double /*t*/, auto, auto scalar) {
+        auto [u, du_dx] = scalar;
+        auto source     = 0.5 * sin(u);
+        auto flux       = du_dx;
+        return serac::tuple{source, flux};
+      },
+      pmesh);
+
+  StdFunctionOperator residual_opr(
+      fes.TrueVSize(),
+      [&x_exact, &residual](const mfem::Vector& x, mfem::Vector& r) {
+        // TODO this copy is required as the sundials solvers do not allow move assignments because of their memory
+        // tracking strategy
+        // See https://github.com/mfem/mfem/issues/3531
+
+        double dummy_time = 0.0;
+
+        const mfem::Vector res = residual(dummy_time, x);
+
+        r = res;
+        r -= residual(dummy_time, x_exact);
+      },
+      [&residual, &J](const mfem::Vector& x) -> mfem::Operator& {
+        double dummy_time = 0.0;
+        auto [val, grad]  = residual(dummy_time, differentiate_wrt(x));
+        J                 = assemble(grad);
+        return *J;
+      });
+
+  const LinearSolverOptions lin_opts = {.linear_solver        = lin_solver,
+                                        .preconditioner       = precond,
+                                        .petsc_preconditioner = pc_type,
+                                        .relative_tol         = 1.0e-10,
+                                        .absolute_tol         = 1.0e-12,
+                                        .max_iterations       = 500,
+                                        .print_level          = 1};
+
+  const NonlinearSolverOptions nonlin_opts = {.nonlin_solver  = nonlin_solver,
+                                              .relative_tol   = 1.0e-10,
+                                              .absolute_tol   = 1.0e-12,
+                                              .max_iterations = 100,
+                                              .print_level    = 1};
+
+  EquationSolver eq_solver(nonlin_opts, lin_opts);
+
+  eq_solver.setOperator(residual_opr);
+
+  eq_solver.solve(x_computed);
+
+  for (int i = 0; i < x_computed.Size(); ++i) {
+    EXPECT_LT(std::abs((x_computed(i) - x_exact(i))) / x_exact(i), 1.0e-6);
```

Comment:
Is there anyway for `x_computed` and `x_exact` to be different sizes at this point?

Reply:
I don't think so, but I can add an assertion for that as well if you want